opnode/rollup/sync/start.go

```diff
@@ -9,102 +40,126 @@ import (
-		nextRefL1, ontoL1, err = reference.RefByL1Num(ctx, refL1.Number+1)
+	reorgDepth := 0
+	// Walk L2 chain from L2 head to first L2 block which has a L1 Parent that is canonical. May walk to L2 genesis
+	for n := l2Head; ; {
+		l1header, err := source.L1NodeByNumber(ctx, n.L1Parent.Number)
 		if err != nil {
-			// If refL1 is the head block, then we might not have a next block to build on the head
-			if errors.Is(err, ethereum.NotFound) {
-				// return the same as the engine head was already built on, no error.
-				nextRefL1 = refL1
-				refL2 = eth.BlockID{Hash: parentL2, Number: refL2.Number}
-				if refL2.Number > 0 {
-					refL2.Number -= 1
-				}
-				err = nil
-				return
+			// Generic error, bail out.
+			if !errors.Is(err, ethereum.NotFound) {
+				return eth.L2Node{}, fmt.Errorf("failed to fetch L1 block %v: %w", n.L1Parent.Number, err)
+			}
+			// L1 block not found, keep walking chain
+		} else {
+			// L1 Block found, check if matches & should keep walking the chain
+			if l1header.Self.Hash == n.L1Parent.Hash {
+				return n, nil
 			}
-			return
 		}
-		// The L1 source might rug us with a reorg between API calls, catch that.
-		if ontoL1 != currentL1 {
-			err = fmt.Errorf("the L1 source reorged, the block for N+1 %s doesn't have the previously fetched block N %s as parent, but builds on %s instead", nextRefL1, currentL1, ontoL1)
+
+		// Don't walk past genesis. If we were at the L2 genesis, but could not find the L1 genesis
+		// pointed to from it, we are on the wrong L1 chain.
+		if n.Self.Hash == genesis.L2.Hash || n.Self.Number == genesis.L2.Number {
+			return eth.L2Node{}, WrongChainErr
 		}
-		return
-	}
 
-	// Search back: linear walk back from engine head. Should only be as deep as the reorg.
-	for refL2.Number > 0 {
-		// remember the canonical L1 block that builds on top of the L1 source block of the L2 parent block.
-		nextRefL1 = currentL1
-		refL1, refL2, parentL2, err = reference.RefByL2Hash(ctx, parentL2, genesis)
+		// Pull L2 parent for next iteration
+		n, err = source.L2NodeByHash(ctx, n.L2Parent.Hash)
 		if err != nil {
-			// TODO: re-attempt look-up, now that we already traversed previous history?
-			err = fmt.Errorf("failed to lookup block %s in L2: %w", refL2, err) // refL2 is previous parentL2
-			return
+			return eth.L2Node{}, fmt.Errorf("failed to fetch L2 block by hash %v: %w", n.L2Parent.Hash, err)
 		}
-		// Check if L1 source has the block that derived the L2 block we are planning to build on
-		currentL1, _, err = reference.RefByL1Num(ctx, refL1.Number)
-		if err != nil {
-			if !errors.Is(err, ethereum.NotFound) {
-				err = fmt.Errorf("failed to lookup block %d in L1: %w", refL1.Number, err)
-				return
-			}
-			// again, if L1 does not have the block, then we just search if we are reorging.
-			err = nil
-			currentL1 = eth.BlockID{} // empty = not found
+		reorgDepth++
+		if reorgDepth >= MaxReorgDepth {
+			panic("Too deep re-org")
```

Comment:
We had a discussion on this before, I did ask for a panic because imho this is an extremely fishy scenario that should never ever happend (after the merge, that is), hence if this gets triggered, human intervention is required is anything that happens past this point is liable to be super wrong.

Returning an error like is done now is not necessarily wrong, but we need to bail at some point, and it's not quite clear to me what is currently being done (there's for sure a retry... so do we just retry forever? or do we drop after some failures We at least need to log something.)

I guess I'm fine with an error as long as we can make sure that the issue can easily be diagnosed in prod.